test/integration-new/cni/eni_config/ipleak_test.go

```diff
@@ -0,0 +1,91 @@
+package env_vars
+
+import (
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/aws/amazon-vpc-cni-k8s/test/framework/resources/k8s/manifest"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	ENI_ENDPOINT = "http://localhost:61679/v1/enis"
+)
+
+var _ = Describe("IP Leak Test", func() {
+	Context("IP Released on Pod Deletion", func() {
+		It("Verify that on Pod Deletion, Warm Pool State is restored", func() {
+			totalIps := 0
+			assignedIps := 0
+			deploymentSpec := manifest.NewBusyBoxDeploymentBuilder().
+				Namespace("default").
+				Name("busybox").
+				NodeName(primaryNode.Name).
+				Replicas(10).
+				Build()
+
+			By("Recording the initial count of IP before new deployment")
+			{
+				totalIps, assignedIps = getTotalAndAssignedIps()
+			}
+
+			By("Deploying a large number of Busybox Deployment")
+			{
+				_, err := f.K8sResourceManagers.
+					DeploymentManager().
+					CreateAndWaitTillDeploymentIsReady(deploymentSpec)
+
+				Expect(err).ToNot(HaveOccurred())
+			}
+
+			By("Recording the count of IP after deployment")
+			{
+				currTotal, currAssigned := getTotalAndAssignedIps()
+				Expect(currTotal).To(Equal(totalIps))
+				// Diff should be equal to number of replicas in new deployment
+				Expect(currAssigned - assignedIps).To(Equal(10))
+			}
+
+			By("Deleting the deployment")
+			{
+				err = f.K8sResourceManagers.DeploymentManager().DeleteAndWaitTillDeploymentIsDeleted(deploymentSpec)
+				Expect(err).NotTo(HaveOccurred())
+			}
+
+			By("Validating that count of IP is same as before")
+			{
+				ip := 0
+				assigned := 0
+				for i := 0; i < 3; i++ {
+					// It takes some time to unassign IP addresses
+					time.Sleep(120 * time.Second)
+					ip, assigned = getTotalAndAssignedIps()
+					if assigned == assignedIps {
+						break
+					}
+				}
+				Expect(ip).To(Equal(totalIps))
+				Expect(assigned).To(Equal(assignedIps))
+			}
+		})
+	})
+})
+
+func getTotalAndAssignedIps() (int, int) {
```

Comment:
I think we should be getting this details from EC2 API instead of getting it from IMDS by doing a pod exec into a pod running with host networking mode. 

We would also prevent using Regex `re := regexp.MustCompile(`\"TotalIPs\":([0-9]*),\"AssignedIPs\":([0-9]*)`)` in that case :)

Reply:
I did try using EC2 api, but these are problems with EC2 api approach 
-> Number of ENI and number of max IP per ENI depends on instance type. So if you want to trigger a creation of new ENI then you need to know how many deployments must be scheduled for that instance type. 
-> The api doesnt tell you which of the IP's are assigned and which of them are free. So IP leak test could not be done. 
Do suggest me ways if I missed some information. I would prefer EC2 api as well.